### PR TITLE
Fixes serialization of ProxyServer for mojo.

### PR DESCRIPTION
**Pull request by mkarolin**
Previuosly, serialization of ProxyServer's HostPortPair used
its ToString method which we overrode, but now
ProxyServerToPacResultElement explicitly constructs the string
by getting host() and port() members and calling
ConstructHostPortString. We need to override this behavior to insert
auth info.

Fixes brave/brave-browser#19371

Chromium change:

https://source.chromium.org/chromium/chromium/src/+/5dbd4599a981da1fac5dc5211772fe843a06a9b0

commit 5dbd4599a981da1fac5dc5211772fe843a06a9b0
Author: Eric Orth <ericorth@chromium.org>
Date:   Fri Sep 24 21:21:24 2021 +0000

    Stop dealing with HostPortPair in proxy_string_util.(cc/h)

    String->ProxyServer now done using ProxyServer::FromSchemeHostAndPort,
    and ProxyServer->String now done using ProxyServer::GetHost() and
    ProxyServer::GetPort(). With less ProxyServer-external dealing with
    HostPortPair, will aide the future-CL transition to not storing
    HostPortPair for all proxies.

    Also has the effect that construction-from-string now results in
    canonicalization of the ProxyServer hostname. Slight impact on tests
    that were parsing stuff with now-recognized-as-invalid hostnames
    (especially the v8 tracing tests that were hacking information into
    wildly invalid hostnames), but don't expect any real-world impact
    because these are clearly invalid names that shouldn't ever be able to
    resolve to real servers.

    Also moved the hostname:port parsing directly into proxy_string_util.cc
    (from url_util::ParseHostAndPort()) to avoid silliness like stripping
    IPv6 brackets just for them to be added back for canonicalization.

    Bug: 1243398

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by darkdh — APPROVED** [review]
We should extend unit test `ProxySpecificationUtilTest.ProxyUriToProxyServer` with username and password cases and run it. So that we can detect brokerage easily in the future. 
It can be done as follow-up so we don't block the release

**Comment by mkarolin** [comment]
@darkdh, added a few unit tests. PTAL. Thanks

**Review by mariospr — APPROVED** [review]
LGTM w/ fixes suggested by @goodov (which mostly apply to code I originally wrote, sorry)

File: chromium_src/net/base/proxy_string_util.cc [hunk 1]
```diff
@@ -94,4 +84,39 @@ std::string ProxyServerToProxyUri(const ProxyServer& proxy_server) {
...
+  std::string proxy_uri = ProxyServerToProxyUri_ChromiumImpl(proxy_server);
+  size_t colon = proxy_uri.find(':');
+  if (colon != std::string::npos && proxy_uri.size() - colon >= 3 &&
+      proxy_uri[colon + 1] == '/' && proxy_uri[colon + 2] == '/') {
+    scheme_prefix = proxy_uri.substr(0, colon + 3);
+    proxy_uri = proxy_uri.substr(colon + 3);  // Skip past the "://"
+  }
+
+  std::string result =
+      scheme_prefix + GetProxyServerAuthString(proxy_server) + proxy_uri;
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply chromium_src/net/base/proxy_string_util.cc]: Fixed in 4339800a25f43be673721686199ff95ae7613e6e

File: chromium_src/net/base/proxy_string_util.cc [hunk 2]
```diff
@@ -94,4 +84,39 @@ std::string ProxyServerToProxyUri(const ProxyServer& proxy_server) {
...
+  std::string proxy_pac =
+      ProxyServerToPacResultElement_ChromiumImpl(proxy_server);
+  size_t space = proxy_pac.find(' ');
+  if (space != std::string::npos && proxy_pac.size() - space >= 1) {
+    scheme_prefix = proxy_pac.substr(0, space + 1);
+    proxy_pac = proxy_pac.substr(space + 1);  // Skip past the space
+  }
+
+  std::string result =
+      scheme_prefix + GetProxyServerAuthString(proxy_server) + proxy_pac;
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply chromium_src/net/base/proxy_string_util.cc]: Fixed in 4339800

File: chromium_src/net/base/brave_proxy_string_util_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "net/base/proxy_string_util.h"
+
+#include "net/base/proxy_server.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace net {
+namespace {
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply chromium_src/net/base/brave_proxy_string_util_unittest.cc]: Fixed in 4339800

File: chromium_src/net/base/proxy_string_util.cc [hunk 4]
```diff
@@ -72,14 +61,15 @@ std::string ProxyServerToProxyUri(const ProxyServer& proxy_server) {
     }
     auth_string += '@';
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply chromium_src/net/base/proxy_string_util.cc]: Fixed in 4339800

**Comment by mkarolin** [comment]
Merging as this is blocking the upcoming release. @goodov, please let me know if any additional fixes are needed and I'll do a followup. Thanks.

**Pull request merged**